design/ceph/object/bucketpolicy.md

```diff
@@ -0,0 +1,93 @@
+
+Followings CRDs are needed to add for defining policies:
+  - _bucketPolicy_ in the `Spec` section of `CephObjectStoreUser`
+  - _bucketPolicy_ in the `parameters` section of `StorageClass` for `ObjectBucketClaim`
+
+Policies need to provide in generic `json` format. A policy can have multiple `statement`.
+
+The following checks are needed to perform in Rook to verify whether the `bucketpolicy` applicable to OBC.
+- for ceph object user, `Principal` value should have username and `Resource` should have specific bucket names. It can be defined for buckets which are not part of the OBC as well, the `bucketname` of an OBC can be fetched from its `configmap`.
+- In `StorageClass`, `Principal` value should be `*`(applicable to all users) and `Resource` should have the bucket name can be empty since it can be generated name from Rook as well and will be attached before setting the policy.
+
+Examples:
+```yaml
+# storageclass-bucket.yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+   name: rook-ceph-delete-bucket
+provisioner: rook-ceph.ceph.rook.io/bucket
+# set the reclaim policy to delete the bucket and all objects
+# when its OBC is deleted.
+reclaimPolicy: Delete
+parameters:
+  objectStoreName: my-store
+  objectStoreNamespace: rook-ceph
+  region: us-east-1
+  bucketName: ceph-bkt
+  bucketPolicy: "Version": "2012-10-17",
+		"Statement": [{
+		"Effect": "Allow",
+		"Principal": {"AWS": ["arn:aws:iam:::*"]},
+		"Action": "s3:PutObjectAcl",
+		"Resource": [
+		"arn:aws:s3:::/*"
+		]
+	}]
+
+# object-user.yaml
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStoreUser
+metadata:
+  name: my-user
+  namespace: rook-ceph
+spec:
+  store: my-store
+  displayName: "my display name"
+  bucketPolicy: "Version": "2012-10-17",
+		"Statement": [{
+		"Effect": "Allow",
+		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
+		"Action": "s3:PutObjectAcl",
+		"Resource": [
+		"arn:aws:s3:::ceph-bkt/*"
+		]
+	}]
+```
+
+The `bucketPolicy` defined on CRD won't override any existing policies on that bucket, will just append. But this can be easily overwritten with help of S3 client, since does not have much control over there.
+
+## Apis and structural changes
+
+The following field will be add to `ObjectStoreUserSpec` and this need to reflected on the existing api's for `CephObjectStoreUser`
+
+```
+type ObjectStoreUserSpec struct {
+        Store string `json:"store,omitempty"`
+        //The display name for the ceph users
+        DisplayName string `json:"displayName,omitempty"`
++       //The list of bucket policy for this user
++       BucketPolicy string `json:"bucketPolicy,omitempty"`
+ }
+```
+
+The `bucket policy` feature is consumed by the brownfield use case of `OBC`, so supporting apis and structures already exists in [policy.go](/pkg/operator/ceph/object/policy.go). Still few more api's are need to read the policy from CRD, validate it and then convert it into `bucketpolicy`, so that can be consumed by existing api's
```

Comment:
The ObjectStoreUser CR isn't used with the OBCs, so when would the buckey policy be applied if specified in the CR? I understand for OBCs it can be applied from the storage class when the bucket is created, but I don't see where the operator can apply the policy from the ObjectStoreUser CR.

Reply:
@travisn Yeah both are different. `bucketpolicy` will be part of the `objectstoreuser` CR itself. At the bucket level both will be applied accordingly. Consider below example : 
```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStoreUser
metadata:
  name: my-user
  namespace: rook-ceph
spec:
  store: my-store
  displayName: "my display name"
  bucketPolicy: "Version": "2012-10-17",
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:PutObject",
		"Resource": [
		"arn:aws:s3:::ceph-bkt-1/*"
		],
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:ListObject",
		"Resource": [
		"arn:aws:s3:::ceph-bkt-2/*"
		],
		"Statement": [{
		"Effect": "Allow",
		"Principal": {"AWS": ["arn:aws:iam:::my-user"]},
		"Action": "s3:GetObject",
		"Resource": [
		"arn:aws:s3:::ceph-bkt-3/*"
		],
	}]
```
In above there are three bucket policies, which need to apply on three different buckets. After creating the user, the policy from CR will be fetched. And on each bucket `putbucketpolicy` will be performed seperately, so we need to make sure that the existing `bucketpolicy`  is not overwritten(just do append)

Reply:
It's not clear to me either, can you describe the complete workflow? Are we saying that for a given `CephObjectStoreUser` every bucket that would be created or already exist will see their policy appended?

Reply:
@leseb : The plan is to set `bucketpolicy` for existing buckets, not for the bucket which is created from this user. The main idea is behind this feature to connect the users to existing buckets and from bucket(or OBC) point of view to provide general access to anonymous users. As of now I didn't consider it for the buckets from  `CephObjectStoreUser`. 

So the flow above cases is that, I have three buckets `ceph-bkt-{1..3}` and I want to give different permission for `my-user`(at this point user may or not exists) on them. After updating the policy in yaml file and when request reaches `reconcileCephUser`, it will divide `bucketpolicy` mentioned in CRD into three different policies. Do `GetBucket` on the bucket one by one, fetch the bucket owner's info, fetch existing policy for each bucket if any and finally append applicable policy  on each bucket calls `setbucketpolicy`.

There is similar flow in [Grant](https://github.com/rook/rook/blob/master/pkg/operator/ceph/object/bucket/provisioner.go#L113) in provisioner.go

Reply:
Thanks for the explanation, if you could add that paragraph about the flow in the design doc.

Reply:
Done